IR TX and RX enhancements

jimmys01 opened the pull request:
Newer IR esp8266 library updated
Updated code to support all the ir devices with the codes given by the IRrecvDumpV2.ino
HVAC support as well.

jimmys01 committed:
IR library added

jimmys01 committed:
Travis build complaining about unused vars

TD-er commented:
The errors (some warnings are treated as errors) : 
```
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'boolean Plugin_035(byte, EventStruct*, String&)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:204:26: error: variable 'IrRepeat' set but not used [-Werror=unused-but-set-variable]
unsigned int IrRepeat=0;
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:205:27: error: variable 'IrSecondCode' set but not used [-Werror=unused-but-set-variable]
unsigned long IrSecondCode=0UL;
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'void parseStringAndSendPronto(String, uint16_t)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:482:34: error: 'newCodeArray' was not declared in this scope
code_array = newCodeArray(count);
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'void parseStringAndSendRaw(String)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:521:33: error: 'newCodeArray' was not declared in this scope
raw_array = newCodeArray(count);
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'void parseStringAndSendGC(String)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:565:34: error: 'newCodeArray' was not declared in this scope
code_array = newCodeArray(count);
^
cc1plus: some warnings being treated as errors
*** [.pioenvs/normal_ESP8266_1024/src/ESPEasy.ino.cpp.o] Error 1
```

You can add a new commit to this branch you used to create this PR. So you don't have to close the PR, just push a new commit to the same branch and it will be appended to this PR. (and I will get a notification)

jimmys01 committed:
commented out more unused code

TD-er commented:
The current errors:
```
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'void parseStringAndSendPronto(String, uint16_t)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:482:34: error: 'newCodeArray' was not declared in this scope
code_array = newCodeArray(count);
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'void parseStringAndSendRaw(String)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:521:33: error: 'newCodeArray' was not declared in this scope
raw_array = newCodeArray(count);
^
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino: In function 'void parseStringAndSendGC(String)':
/home/travis/build/letscontrolit/ESPEasy/src/_P035_IRTX.ino:565:34: error: 'newCodeArray' was not declared in this scope
code_array = newCodeArray(count);
^
```

jimmys01 commented:
Those errors are false negatives. The project compiles in adruino ide without a hick

TD-er commented:
@jimmys01 It is being built in PlatformIO, so even if it seems to build in Arduino IDE at your computer, if it fails by Travis, it will fail when I merge it.
And those errors are not just warnings that are being reported as error, it is a true compiler error.

jimmys01 commented:
The part of the code in question works in production. I do not know why it failed. if someone has any idea please let me know. Arduino IDE doesn't complain

svmac commented:
jimmys01 I think you should commit together the library upgrade IRRemoteESP8266 and your original PR

jimmys01 committed:
Travis please be happy :-)

TD-er commented:
This library upgrade does add a lot to the resource usage of ESPeasy.
I just built the "test" version here and it adds at least 15k to the firmware size and over 4k of memory use.

Since it is currently included in the "normal" build, it will have effect on all users.

So my suggestion is to move this plugin to a separate build, since it only has very specific use case.
What other set of plugins should be included in that build? The "normal" set?
When using the plugin set of "test" you now have less than 10 k free at max and the sketch will be 814k in size, so that is really a lot.

Just 2 builds for testing purposes:
[TestBuild_ESPeasy_PR1645](https://www.dropbox.com/s/ugnsj9snh6oh7gk/TestBuild_ESPeasy_PR1645.rar?dl=0)

jimmys01 commented:
There is also the RX plugin that can be checked how much is costing in resources and if it can be omited.
Also we can also make a build seperate for HVAC devices that they also contribute a lot to the size of the build

TD-er commented:
The plugin itself is not really adding to the resource consumption.
It is the library.
And what plugins do you mean for HVAC?

jimmys01 commented:
The library is writen with (example) #if SEND_TOSHIBA_AC and with #if DECODE_TOSHIBA_AC all over so we can decide what to include and what not.   For example we can choose not to include the HVAC protocols in one build and include them in another.

Also if you check the IRremoteESP8266.h it sais 
> // Each protocol you include costs memory and,  during decode, costs time
> // Disable (set to false) all the protocols you do not need/want!
> // The Air Conditioner protocols are the most expensive memory-wise.

TD-er commented:
Ah, now I get it :)

TD-er committed:
[IR/IRTX] For PullReq #1645 split IR/IRTX to a separate build

The new library takes a lot of resources.

TD-er commented:
Ah, it looks like I can even add a commit to an existing pull request of someone else.
That's nice :)
I added a commit to make a separate build for the IR parts.
If you could test that, I can merge it tomorrow, or as soon as you've tested it.

jimmys01 committed:
IR RX updated

The IRrecvDumpV2.ino functionality added to the RX module.
Let me know for a good way to display the RAW values.

jimmys01 commented:
Go ahead and merge it :-)

This is a preview of the decoding log 
What is those ^ characters doing in my log? 
```
27307: IR: Encoding : GREE^Code : 0108205000200050 (64 bits)^
27308: Power: Off, Mode: 1 (COOL), Temp: 24C, Fan: 0 (AUTO), Turbo: Off, XFan: Off, Light: On, Sleep: Off, Swing Vertical Mode: Manual
28982: IR: Encoding : UNKNOWN^Code : EAF62E02 (69 bits)^
31306: IR: Encoding : GREE^Code : 19092050122000F0 (64 bits)^
31307: Power: Off, Mode: 1 (COOL), Temp: 25C, Fan: 1, Turbo: Off, XFan: Off, Light: On, Sleep: Off, Swing Vertical Mode: Manual, Swing
33006: IR: Encoding : GREE^Code : 19092050122000F0 (64 bits)^
33006: Power: Off, Mode: 1 (COOL), Temp: 25C, Fan: 1, Turbo: Off, XFan: Off, Light: On, Sleep: Off, Swing Vertical Mode: Manual, Swing
52982: IR: Encoding : NEC^Code : 20DF23DC (32 bits)^
```

TD-er commented:
@jimmys01 
This part of the log line generation is a bit strange:

```C++
         // If we got a human-readable description of the message, display it.
          if (log != "")  log += ("Mesg Desc.: " + log);
```

This should probably be "=" instead of "+="
Also use the F-macro, since this text will be stored in memory.
Still that doesn't explain the '^' in the logs.

jimmys01 committed:
Small correction

jimmys01 commented:
ok those ^ are supposed to be line changes

crankyoldgit commented:
@jimmys01 FYI, while technically not required by the license (I think, not sure :), it would be polite to at least reference where you _borrowed_ a lot of the code for your additions to `_P035_IRTX.ino`.
e.g. https://github.com/letscontrolit/ESPEasy/blob/mega/src/_P035_IRTX.ino#L278 is very similar to https://github.com/markszabo/IRremoteESP8266/blob/master/examples/IRMQTTServer/IRMQTTServer.ino#L461

No objections to letscontrolit etal using the code, but an attribution/reference would be nice. ;-) It may help maintainers know where to look upstream to find fixes for bugs etc later on.

TD-er commented:
@jimmys01 Did you use the code from there, or is there also another source using the same code?
At least a single [Google search](https://www.google.se/search?rlz=1CAACAV_enNL785NL786&ei=bQqgW7SjNuKorgSjpZigDQ&q=%22Parse+an+Air+Conditioner+A%2FC+Hex+String%2Fcode+and+send+it.%22&oq=%22Parse+an+Air+Conditioner+A%2FC+Hex+String%2Fcode+and+send+it.%22&gs_l=psy-ab.3...5598.9584.0.10041.2.2.0.0.0.0.157.213.1j1.2.0....0...1.1.64.psy-ab..0.0.0....0.dBUliF6UwEA) shows a lot of hits on exactly that quote (not searched yet for other lines which are the same)
I agree the resemblance is striking  (perhaps the diff is even 0), so a request for credits seems to be in order.
I will add it when I get back from Sweden. (please help me remind, or someone else may add a PR to add a reference)
What should be mentioned, "David Conran" or @crankyoldgit  ?

crankyoldgit commented:
@TD-er Either is fine. Go with what ever is the norm in your code base, or even a generalised "Some of this code was borrowed from the IRremoteESP8266 library" or similar is fine.

FYI, most of the initial hits on that search are copies of my code in some form or other.
**NSFW warning**: For what ever reason, some of those links on that search query go to porn sites. I found out the hard way. I have no idea why they chose my code/comment for that. [Rule 34 I guess.](https://www.urbandictionary.com/define.php?term=Rule%2034) :-/

jimmys01 commented:
All the HVAC routines are from the example scetches and the irdumpv2.ino.

Στις 18 Σεπ 2018 01:28, ο χρήστης David Conran <notifications@github.com> έγραψε:

@TD-er<https://github.com/TD-er> Either is fine. Go with what ever is the norm in your code base, or even a generalised "Some of this code was borrowed from the IRremoteESP8266 library" or similar is fine.

FYI, most of the initial hits on that search are copies of my code in some form or other.
NSFW warning: For what ever reason, some of those links on that search query go to porn sites. I found out the hard way. I have no idea why they chose my code/comment for that. Rule 34 I guess.<https://www.urbandictionary.com/define.php?term=Rule%2034> :-/

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/letscontrolit/ESPEasy/pull/1645#issuecomment-422192528>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AQ7RY3mV0w0QI31ay3O9anmwZmzkpnrjks5ucCIAgaJpZM4V6Kim>.
